pkg/spdk/engine.go

```diff
@@ -1229,6 +1310,410 @@ func (e *Engine) validateAndUpdateReplicaNvme(replicaName string, bdev *spdktype
 	return e.ReplicaStatusMap[replicaName].Mode, nil
 }
 
+// Expand performs an online volume expansion for the Longhorn Engine using SPDK.
+// It expands the underlying replica logical volumes (lvol), recreates the SPDK RAID bdev,
+// suspends and resumes frontend I/O as needed, and ensures cleanup and status updates on failure.
+func (e *Engine) Expand(spdkClient *spdkclient.Client, size uint64, superiorPortAllocator *commonbitmap.Bitmap) (retErr error) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.log.Info("Expanding engine")
+
+	// fetch all replica clients
+	replicaClients, err := e.getReplicaClients()
+	if err != nil {
+		return err
+	}
+	defer e.closeReplicaClients(replicaClients)
+
+	// requireExpansion checks if the expansion is needed and validates the request.
+	requireExpansion, err := e.requireExpansion(size, replicaClients)
+	if err != nil {
+		return errors.Wrap(err, "requireExpansion failed")
+	}
+
+	// engineErr will be set when the engine failed to do any non-recoverable operations.
+	var (
+		expanded = false
+	)
+	defer func() {
+		e.finishExpansion(expanded, size, retErr)
+	}()
+
+	if !requireExpansion {
+		e.log.Info("No need to expand engine")
+		expanded = true
+		return nil
+	}
+
+	// prepareRaidForExpansion checks if RAID exists, suspends frontend if needed, and deletes the RAID bdev.
+	isSuspended, bdevRaidUUID, err := e.prepareRaidForExpansion(spdkClient)
+	if isSuspended {
+		defer func() {
+			if frontendErr := e.initiator.Resume(); frontendErr != nil {
+				retErr = multierr.Append(retErr, errors.Wrapf(frontendErr, "original error; resume failed"))
+			}
+		}()
+	}
+	if err != nil {
+		return errors.Wrap(err, "prepare raid for expansion failed")
+	}
+
+	// expand replicas
+	if err := e.expandReplicas(replicaClients, spdkClient, size); err != nil {
+		return errors.Wrap(err, "expand replica failed")
+	}
+
+	// recreate RAID and reconnect frontend
+	if err := e.reconnectFrontend(spdkClient, bdevRaidUUID, superiorPortAllocator); err != nil {
+		return errors.Wrap(err, "reconnectFrontend failed")
+	}
+	e.log.Info("Expanding engine complete")
+
+	expanded = true // which could be true even in partial success
+	return nil
+}
+
+func (e *Engine) requireExpansion(size uint64, replicaClients map[string]*client.SPDKClient) (requireExpansion bool, err error) {
+	if e.isExpanding {
+		return false, fmt.Errorf("expansion is in progress")
+	}
+
+	if e.IsRestoring {
+		return false, fmt.Errorf("restoring is in progress")
+	}
+
+	if e.SpecSize > size {
+		return false, fmt.Errorf("cannot expand engine to a smaller size %v, current size %v", size, e.SpecSize)
+	}
+	if e.SpecSize == size {
+		e.log.Infof("Engine already at requested size %v, skipping expansion", size)
+		return false, nil // no need to expand
+	}
+
+	roundedNewSize := util.RoundUp(size, helpertypes.MiB)
+	if roundedNewSize != size {
+		return false, fmt.Errorf("rounded up spec size from %v to %v since the spec size should be multiple of MiB", size, roundedNewSize)
+	}
```

Comment:
Same question in replica.go.
Create an improvement to move the check to longhorn-manager.